src/world/map.c

```diff
@@ -1709,10 +1970,258 @@ static money32 map_set_land_height(int flags, int x, int y, int height, int styl
 			return MONEY32_UNDEFINED;
 		}
 	}
-
-	int eax = x, ebx = flags, ecx = y, edx = (style << 8) | height, esi = 0, edi = selectionType << 5, ebp = 0;
-	RCT2_CALLFUNC_X(0x006639FE, &eax, &ebx, &ecx, &edx, &esi, &edi, &ebp);
+	
+	/*	Continuing dissassembly - work in progress
+	 *	Known bugs:
+	 *		Raising the corner of a piece of land doesn't remove walls.
+	 *		Small scenery elements (ones that take up 1/4 of a tile) are not removed when changing land height.
+	 *	Original bugs:
+	 *		Lowering the land tile underneath a wall removes the wall.
+	 *		Land can be raised over flat scenery items (gardens, for example) without removing them.
+	 */
+	
+	int eax = x, ebx = flags, ecx = y, edi = selectionType << 5, ebp = 0;
+	uint32 edx = (style << 8) | height;
+	uint32 esi = 0;
+	
+	//Uncomment to use vanilla code
+	//RCT2_CALLFUNC_X(0x006639FE, &eax, &ebx, &ecx, (int *)&edx, (int *)&esi, &edi, &ebp); return ebx;
+	
+	RCT2_GLOBAL(0x9E2E18, uint32) = 0;
+	if(ebx&GAME_COMMAND_FLAG_APPLY)
+	{
+		footpath_remove_litter(eax, ebx, map_element_height(eax, ecx));
+		//Hmmm... I have a bug in here where raising the corner of a tile doesn't remove walls on it.
+		map_remove_walls_at(eax, ecx, edx*8-16, edx*8+32);
+	}
+	
+	RCT2_GLOBAL(0x9E2E18, uint32) += 0xC8;
+	
+	//RCT2_CALLFUNC_X(0x663A3A, &eax, &ebx, &ecx, (int *)&edx, (int *)&esi, &edi, &ebp); return ebx;
+	
+	int saved_eax = eax;
+	int saved_ecx = ecx;
+	unsigned int saved_edx = edx;
+	
+	uint8 *pdl = (uint8 *)&edx;
+	uint8 *pdh = (uint8 *)&edx+1;
+	uint8 *pbl = (uint8 *)&ebx;
+	uint8 *pbh = (uint8 *)&ebx+1;
+	
+	/*
+	esi = ecx&0xFFFF; //movzx esi, cx
+	esi <<= 8;
+	esi |= eax;
+	esi >>= 3;
+	esi = RCT2_ADDRESS(RCT2_ADDRESS_TILE_MAP_ELEMENT_POINTERS, uint32)[(int)esi/4];
+	*/
+	esi = (int)map_get_first_element_at(eax/32, ecx/32);
```

Comment:
Is there any reason why you changed esi from type `rct_map_element*` to an int. It is much easier to follow if you keep it as a `rct_map_element`